Summer 2020 Update

Daehoon7992 opened the pull request:
1. removed library(rlocker)

2. reformat the app

3. add references

4. changed the button size to large

5. changed relative risk formula

Daehoon7992 committed:
6 updates

1. removed library(rlocker)

2. reformat the app

3. add references

4. changed the button size to large

5. changed relative risk formula

Daehoon7992 committed:
Two updates

1. changed the position of 'next question' button and changed the name to 'next>>'

2. changed some questions on question banks where there were some wording issues in the App.

Daehoon7992 committed:
Two more updates - libraries

removed 'rlocker' library and related codes in server.r

add reference for 'plotrix'

neilhatfield reviewed (changes requested):
Here are some items that need to be addressed:

1. I believe that the packages `shinyjs`, `shinyWidgets`, and `plotrix` can be removed from both the ui and server files. (Be sure to thoroughly test and then update the references.)
2. In the UI,
    - Remove excess white space (empty lines). (A single empty line can be used to separate chunks of code, but try to leave comments instead.)
    - For lines that go beyond 80 characters, break the lines with returns so that the line moves down. This will improve readability. (RStudio puts a vertical greyline at the 80 character mark in the editor.
    - Update the instructions to reflect that the game is about a guy falling out of a tree rather than traditional hangman (i.e., remove the gallows references).
    - Delete the `class = "circle grow"` from the button definition (i.e., line 76)
    - Remove the `br()` in the Acknowledgements to bring the two sentences in to a single paragraph.
    - Remove the `h2` tags from all of the list items. (Only "Prerequisites" needs to be in h2.)
    - Replace the definition of probability with the following:
        > Probability is the long-run relative frequency of a particular data even occurring, given our assumptions about a repeatable process. For example, if we were to imagine running a lottery to pick a US citizen, the probability of pecking a female is 50.8%. Thus, 50.8% of the time we repeat carrying out this lottery (always starting from the same initial population), we will pick a female. (Note: we won't pick the exact same person each time.)
    - Replace the Risk & Relative Risk definition with the following: (note: break Risk and Relative Risk into separate list items.
        > Risk refers to the probability of a data event that we view as being negative, undesirable, or "bad". For example, the lifetime risk of developing skin cancer is about 2.6% (1/38) for white Americans. Thus, we if imagine a process of picking white Americans and observing their lifetimes, 2.6% of the time we will pick an individual who will develop skin cancer. 
    - For Relative Risk,
        > Relative Risk (RR) is the ratio of two groups' risk (probability) for a particular data event. We often calculate relative risk through the formula [insert the code for the formula here] For example, the relative risk of skin cancer for a white American is \\(2.6\%/0.1\% = 0.0026/0.001 = 26\\) times as large as Black Americans.
    - Odds and Odds Ratios (remove the colon)
        > Odds, often expressed as a ratio called an Odds Ratio, compares the probability a data event happen with the probability of the same event not happen (i.e., the complement). The odds ratio of a white American getting skin cancer is \\(\\frac{2.6}{100-2.6}\\approx 0.0267\\). Thus, the probability (risk) of a white American getting skin cancer is 0.0267 times as large as the probability of them not getting skin cancer. Put another way, we can say that the odds of a white American getting skin cancer is 1 to 37. The odds of a black American getting skin cancer are 1 to 999.
    - Skin Cancer data is from https://www.cancer.org/cancer/melanoma-skin-cancer/about/key-statistics.html, please add to references.
    - Change the Challenge tab to Game and update the icon used.
    - Remove the in-line css (lines 150-151).
    - Move the question text out of the well panel across the top.
    - What is the sidebar layout section doing? Shift this to a fluid row with two columns. In the first (left) column, use a well panel for the matches. For the second (right) column, place the score and picture. Play around with the column widths to find ones that work well.
    - The instructions at the top of the choices box should not be in `h3`; move to `p`.
    - For the selectInputs (lines 173-232), play around with the width parameter to find an option that doesn't wrap as they currently do. You might start out by removing the `width` argument completely and seeing how that looks.
    - Remove lines 261-263.
    - Why is there something called "distPlot"? That doesn't match our naming conventions.
    - All of the selectInputs have empty labels which is an accessibility issue. To resolve this, what if we label the boxes with Probability, Risk, Relative Risk, etc. (you'll have to add a fifth) and make the choices the numeric values plus an option for unused (we'll have to play with the wording). This will resolve the accessibility issues with the current app and improve what the user is supposed to do.
    - In the Next button, put a space between "Next" and ">>".
3. In the server,
    - Add `library(boastUtils)`.
    - Remove excess white space (empty lines) (See above)
    - Remove any commented out code that no longer serves a purpose (e.g., lines 39-41)
    - Create an issue on the master branch that notes that the server code needs to be examined and and made flexible. Mention that the code is linear with excessive if-else usages and the question bank is hard coded. (If you want to try tackling this, you may but don't feel that you have to this summer.)
4. In the www folder, remove/delete the following files:
    - fail.jpg and fail.png
    - homebut.PNG
    - logo.png
    - pass.jpg
    - rr.png
    - any files ending in CSS
5. There are some contrast error currently in the app. This is not unexpected and deals with the central BOAST CSS file. You do not need to worry about these. However, if you have suggestions for tweaking the Yellow theme, please don't hesitate to send them to me.

Daehoon7992 committed:
6 updates

1. Challenge -> Game
2.  wording change on 'instruction'
3. changed some UI on 'game tab'
4. changed some wordings on question bank
5. add re-attempt button
6. add 'alert' feature when the user gets 10 answers correct

Daehoon7992 committed:
9 updates

1. deleted files from www folder
2. deleted 'shinyjs' and 'plotrix' library
3. revised wordings 'Neil`s request above.
4. removed 'class = 'cirle grow''
5. replace sidebarLayout to fluidRow
6. add reference for 'skin caner'
7. removed excess white lines
8. re-able reset button. So, when the user finishes, the user can reset the app.
9. add another 'Alert' for failing.
So, when the user fails all chances, the game is over.

Daehoon7992 committed:
5 updates

1.  add 'observeEvent(input$end,~)', so that when user gets all 10 scenarios correct, the user goes to the reference page.

2. changed wording from 'Number of correct answers:' to 'Number of completed scenarios:'.

3. update some buttons depends on the situations.

4. changed the picture size to fit, so that it fits on every device.

5. add some instructions because of the step updates.

Daehoon7992 committed:
Minor and final(?) updates

1. Add rlocker library again for future work

2. Changed some wordings on 'Acknowledgement'.

Daehoon7992 committed:
updates on quesion bank and new stuff 'Increased Risk'

1. Fixed some words on the question bank.

2. Added i button on the Header

3. Added 'Increased Risk' on pre-req and game tab.

neilhatfield reviewed (changes requested):
Here are the next set of changes:

1. In the questionbank.csv,
    - I would replace all of the 1), 2), 3), 4) with A), B), C), D)
    - I'm noticing some strange characters in rows 6-9 ("A¬†¬ìhull loss¬î¬†is an"). While these are getting rendered in the app, I would like to you remove these characters just as a precaution.
    - I'm seeing the ¬ symbol in other lines too; fix.
    - Capitalize "apnea" in "Obstructive Sleep Apena"
    - There seems to be extra space in "National Highway Traffic Safety Administration"; remove.
2. In the UI,
    - You can remove the comment in line 4
    - You can remove line 6 (rlocker gets loaded by boastUtils)
    - Note: in line 7 you've used the wrong symbol for the apostrophe. This doesn't matter since this is a commented line, but just be careful as this could cause problems outside of a comment.
    - Lines 44-45: these are long and extend past the vertical gray line that denotes a long line. Try to break these lines to improve readability of your code.
        - Lines 35-36 are a good example of breaking long lines. The href in line 36 is a case that should *__not__* be broken (just as you've done).
        - Deal with long lines everywhere in your code.
    - Are there any two-by-two tables in the app?
    - Add a comma between "odds" and "and odds ratio" in the first paragraph.
    - Instruction modifications:
        - Line 1: stays as is.
        - Line 2: "You'll start the game with a man relaxing on top of a tree and a scenario displayed."
        - Line 3:  "For each scenario, you'll need to correctly identify which type of measure is each number that gets displayed. Press the Submit button to check your work."
        - Line 4: "If you misidentify any values, the man will fall down to the next branch. If you correctly identify all of the values, you'll have completed that scenario."
        - Line 5: "Your goal is to complete 10 scenarios before the man falls to the ground. The man will fall to the ground after four mistakes."
    - As a paragraph between the list of instructions and the Go button:
        > Note: if you misidentify all of the values for a scenario, you won't be able to move beyond that scenario until you correctly identify at least one value. Use the Re-attempt button to make a new attempt at identifying the values in the current scenario. Press the RESET button if you wish to start the game over.
    - Increased Risk
        - Move the formula above the example.
        - There needs to be a comma after "For example"
        - Make the example match the relative risk example (i.e., replace the wifi signal with the skin cancer) (I want a consistent example theme to help highlight the differences and relationships between these measures; jumping into different scenarios is counterproductive.)
    - Odds
          > Odds is the ratio of two probabilities--the probability of a data event and the probability of that data event not happening (i.e., the complement or the opposite event). There are two ways that odds are expressed: as a fraction or using 'odds notation' with a colon. Letting \\(p=\frac{x}{N}\\) represent the probability of a data event happening \\[\\text{Odds}=\\frac{p}{1-p}=\\frac{x/N}{(N-x)/N}\\equiv X:(N-X)\\] We read odds notation, \\(X:Y\\) as "X to Y".
            - Don't forget to use double backslashes where the single ones are.
        - Replace the horse racing example with the odds of a person getting skin cancer.
    - Odds Ratio
        > An odds ratio is the ratio of the odds for two groups of the same/similar data event. Suppose that the first group has a probability of the data event of p and the second group has a probability of r. Then \\[\\text{Odds Ratio}=\frac{p}{1-p}\\bigg/\frac{r}{1-r}\\]
        - Replace the heart attack example with skin cancer for white and black Americans.
3. This comment will deal with the uiOutput("box#") in the UI and the output$box# in the server
    - As I examine more of what is going on this app and try to track down accessibility issues, I'm getting a better understanding of what was originally done in the app.
    - As things are now, the drop downs are currently label-less (i.e., label = ''). This is an accessibility violation and must be addressed.
    - My recommendation is to 1) ditch the uiOutput('box#'), 2) create a default label = 'first value to identify' for each of the drop downs, and 3) re-code the server section to make use of updateSelectInput (in particular using the label argument to pass a new label to the UI.)
4. In the server,
    - See prior comment about long lines
    - Feel free to change the names of objects so that the names are more informative.
    - Make sure to be explicit in the code (i.e., list the arguments by name)
    - In lines 149 and 153, there is an extra set of quotation marks denoting an additional (and empty) string for the paragraph. Remove this. In line 149, add a space between the colon and the close quotation mark.
    - You've already listed the hard coding issue (#3). Add a note that the hard coding starts with the Questions section circa line 156.
    - Add a space after commas in your code. For example, line 401: add a space between the comma and the width argument.

Daehoon7992 committed:
label updates on selectInput and instuctions update

1. Instuctions updated

2. Added labl on selectInputs from nothing to input$box#

neilhatfield committed:
Updates

I've updated the files in several places and left comments for you to go through and app

neilhatfield reviewed (changes requested):
I've pushed some changes to the files:

1. I fixed a small typo in the question bank
2. I've made multiple changes to the UI and server files. 
    - I have left you comments in each for you to look through and finish carrying out the process. If you search on your name (Daehoon), you'll be able to locate them.
    - When you make the changes, be sure to delete the unneeded code.

Daehoon7992 committed:
Updates after 07/17`s feedback

UI.r
1) Updated the inputID for readability

2) changed the imageOutput to uiOutput

3) fixed 'Increase' to 'Increased'

4) basically updated selectInput from the guidance that professor Neil gave me

Server.r
1) merge the observeEvent functions into one for 'submit', 'reset', 'nextq', and 'reattempt'

2) Added the 'gameProgess', so that the user can continue their game whenver they exit from gam tab to prereq page and come back. - Thanks for 'Hypothesis Testing Game' app.

3) Removed unnecessary 'correct_answer <- as.matrix(bank[1:60,1])' in observeEvent functions

4) Removed the HACK 'input$tabName', and split that into 'intput$go2' and 'input$tabs'

5) Changed the hardcode for question output into one line - thanks to Neil`s idea about the calculation

6) Added 'observe' for updateSelectInput, so that it automatically updates everytime the scenario changes. This also helps to fix previous label = '' problem

** Though I still need to add updateSelectInput function on 'reattempt', and 'reset' observeEvent. I guess it is because those observeEvent doesn`t update value$index(which is for the scenario)

7) The user now reset the game once the user gets all 10 correct scenarios

8) Leave imageOutput for 'scoreTree', because it just looks fine.

9) Added comments every line, except for repetitive ones

neilhatfield reviewed (dismissed):
Nicely done.

I'll be making a few minor tweaks and then I'll merge this into the Master Branch.

neilhatfield committed:
Update ui.R

Removed an unused library and added the licensing info.
